grails-app/services/org/pih/warehouse/inventory/StockMovementService.groovy

```diff
@@ -1083,12 +1099,117 @@ class StockMovementService {
         List<PackPageItem> packPageItems = new ArrayList<PackPageItem>(items)
 
         if (max != null && offset != null) {
-            return packPageItems.subList(offset.toInteger(), offset.toInteger() + max.toInteger() > packPageItems.size() ? packPageItems.size() : offset.toInteger() + max.toInteger())
+            return packPageItems.subList(offset, offset + max > packPageItems.size() ? packPageItems.size() : offset + max)
         }
 
         return packPageItems
     }
 
+    List<ImportPickCommand> parsePickCsvTemplateImport(String text) {
+        try {
+            def csvMapReader = new CSVMapReader(new StringReader(text), [skipLines: 1])
+            csvMapReader.fieldKeys = [
+                    'id',
+                    'code',
+                    'name',
+                    'lot',
+                    'expiration',
+                    'binLocation',
+                    'quantity',
+            ]
+            List<Map> dataList = csvMapReader.toList()
+            return dataList*.asType(ImportPickCommand)
+
+        } catch (Exception e) {
+            throw new RuntimeException("Error parsing order item CSV: " + e.message, e)
+        }
+    }
+
+    void validatePicklistListImport(List<PickPageItem> pickPageItems, Boolean supportsOverPick, List<ImportPickCommand> picklistItems) {
+        Map<String, List> groupedItems = picklistItems.groupBy { it.id }
+
+        picklistItems?.each { ImportPickCommand data ->
+            data.validate()
```

Comment:
I'm actually struggling with understanding what this validate()'s role is here - we just call it "in the air", but I don't see you relying on it anywhere below, since still you proceed with some kind of manual validation?

Reply:
I need to call validate before calling other `rejectValue()` because if I first do my manual validation with rejectValue and then call validate it will clear all of the existing errors and only generate ones which failed dues to constraints specified in the Command file.

Later in the import method I check if command record has any errors and skip the import of that row.